File: android/app/src/test/java/com/nextpage/data/remote/addons/AddonRegistryTest.kt
package com.nextpage.data.remote.addons

import com.nextpage.data.local.dao.AddonDao
import com.nextpage.data.local.entity.AddonEntity
import com.nextpage.data.remote.catalog.CatalogErrorCode
import com.nextpage.data.remote.catalog.CatalogProvider
import com.nextpage.data.remote.catalog.CatalogSourceInfo
import com.nextpage.data.remote.catalog.CatalogSourceKind
import com.nextpage.data.remote.catalog.CatalogException
import com.nextpage.data.remote.catalog.PagedResult
import com.nextpage.data.remote.catalog.addonSource
import kotlinx.coroutines.test.runTest
import org.json.JSONArray
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

private val VALID_MANIFEST_JSON = """
    {
      "id": "my-addon",
      "name": "My Addon",
      "version": "1.0.0",
      "catalogs": [{ "type": "book", "id": "main", "name": "Main catalog" }],
      "resources": ["catalog"]
    }
""".trimIndent()

private fun manifestJson(id: String = "my-addon", name: String = "My Addon", version: String = "1.0.0"): String =
    JSONObject()
        .put("id", id)
        .put("name", name)
        .put("version", version)
        .put(
            "catalogs",
            JSONArray().put(JSONObject().put("type", "book").put("id", "main").put("name", "Main catalog"))
        )
        .put("resources", JSONArray().put("catalog"))
        .toString()

private fun resource(
    json: String? = null,
    name: String = "My Addon",
    version: String = "1.0.0",
    status: Int = 200,
    contentType: String? = "application/json"
) = AddonResource(
    status,
    contentType,
    (json ?: manifestJson(name = name, version = version)).toByteArray(Charsets.UTF_8)
)

/** Transport returning canned resources per call, in order (last one repeats). */
private class ScriptedAddonTransport(vararg resources: AddonResource) : AddonHttpTransport {
    private val queue = resources.toMutableList()
    var calls = 0
        private set

    override suspend fun fetch(url: String): AddonResource {
        calls += 1
        return if (queue.size > 1) queue.removeAt(0) else queue[0]
    }
}

/** In-memory mirror of the Room AddonDao UPSERT semantics (reinstall preserves enabled). */
private class FakeAddonDao : AddonDao {
    val rows = mutableListOf<AddonEntity>()

    override suspend fun getAll(): List<AddonEntity> = rows.toList()
    override suspend fun getById(id: String): AddonEntity? = rows.firstOrNull { it.id == id }
    override suspend fun getByUrl(url: String): AddonEntity? = rows.firstOrNull { it.url == url }
    override suspend fun upsert(addon: AddonEntity) {
        val i = rows.indexOfFirst { it.id == addon.id }
        if (i >= 0) rows[i] = addon else rows.add(addon)
    }

    override suspend fun setEnabled(id: String, enabled: Boolean) {
        val i = rows.indexOfFirst { it.id == id }
        if (i >= 0) rows[i] = rows[i].copy(enabled = enabled)
    }

    override suspend fun delete(id: String) {
        rows.removeAll { it.id == id }
    }
}

private class FakeCatalogProvider(name: String, kind: CatalogSourceKind) : CatalogProvider {
    private val source = CatalogSourceInfo("fake:$name", name, kind)
    override fun listSources(): List<CatalogSourceInfo> = listOf(source)
    override suspend fun search(query: String, page: Int): PagedResult = PagedResult(emptyList(), null, 0)
    override suspend fun getDetails(id: String): Nothing = throw CatalogException(CatalogErrorCode.NOT_FOUND, id)
    override fun resolveDownloadUrl(formats: Map<String, String>, preferEpub: Boolean): String = ""
}

class AddonRegistryTest {

    private fun registry(dao: FakeAddonDao, vararg resources: AddonResource): AddonRegistry =
        AddonRegistry(dao, ScriptedAddonTransport(*resources))

    @Test
    fun `install stores validated manifest keyed by sha256 addonId`() = runTest {
        val dao = FakeAddonDao()
        val url = "https://example.com/manifest.json"
        val manifest = registry(dao, resource()).install(url)
        assertEquals("my-addon", manifest.id)
        assertEquals(1, dao.rows.size)
        assertEquals(AddonId.fromUrl(url), dao.rows[0].id)
        assertEquals(url, dao.rows[0].url)
        assertTrue(dao.rows[0].enabled)
        assertEquals("My Addon", JSONObject(dao.rows[0].manifestJson).getString("name"))
    }

    @Test
    fun `install rejects non-https urls before any network io`() = runTest {
        val dao = FakeAddonDao()
        val transport = ScriptedAddonTransport(resource())
        val reg = AddonRegistry(dao, transport)
        for (url in listOf("http://example.com/manifest.json", "ftp://example.com/manifest.json")) {
            try {
                reg.install(url)
                fail("expected HTTPS_REQUIRED for $url")
            } catch (err: AddonFetchException) {
                assertEquals(AddonFetchErrorCode.HTTPS_REQUIRED, err.code)
            }
        }
        assertEquals(0, transport.calls)
        assertEquals(0, dao.rows.size)
    }

    @Test
    fun `reinstall is idempotent - manifest updated, enabled preserved`() = runTest {
        val dao = FakeAddonDao()
        val url = "https://example.com/manifest.json"
        val reg = registry(dao, resource(), resource(version = "2.0.0"))
        reg.install(url)
        val id = dao.rows[0].id
        reg.setEnabled(id, false)
        reg.install(url)
        assertEquals(1, dao.rows.size)
        assertEquals(id, dao.rows[0].id)
        assertEquals("2.0.0", JSONObject(dao.rows[0].manifestJson).getString("version"))
        assertFalse(dao.rows[0].enabled)
    }

    @Test
    fun `install rejects non-2xx status as network error`() = runTest {
        val dao = FakeAddonDao()
        try {
            registry(dao, resource(status = 404)).install("https://example.com/manifest.json")
            fail("expected NETWORK")
        } catch (err: AddonFetchException) {
            assertEquals(AddonFetchErrorCode.NETWORK, err.code)
        }
        assertEquals(0, dao.rows.size)
    }

    @Test
    fun `install rejects html content type`() = runTest {
        val dao = FakeAddonDao()
        try {
            registry(dao, resource(json = "<html>not a manifest</html>", contentType = "text/html"))
                .install("https://example.com/manifest.json")
            fail("expected BAD_CONTENT_TYPE")
        } catch (err: AddonFetchException) {
            assertEquals(AddonFetchErrorCode.BAD_CONTENT_TYPE, err.code)
        }
        assertEquals(0, dao.rows.size)
    }

    @Test
    fun `install rejects oversized manifests before parse`() = runTest {
        val dao = FakeAddonDao()
        try {
            registry(dao, AddonResource(200, "application/json", ByteArray(65 * 1024)))
                .install("https://example.com/manifest.json")
            fail("expected TOO_LARGE")
        } catch (err: AddonFetchException) {
            assertEquals(AddonFetchErrorCode.TOO_LARGE, err.code)
        }
        assertEquals(0, dao.rows.size)
    }

    @Test
    fun `install rejects manifest id colliding with built-in source name`() = runTest {
        val dao = FakeAddonDao()
        try {
            registry(dao, resource(json = manifestJson(id = "gutendex")))
                .install("https://example.com/manifest.json")
            fail("expected INVALID_MANIFEST")
        } catch (err: AddonFetchException) {
            assertEquals(AddonFetchErrorCode.INVALID_MANIFEST, err.code)
        }
        assertEquals(0, dao.rows.size)
    }

    @Test
    fun `uninstall removes the row - other rows unaffected`() = runTest {
        val dao = FakeAddonDao()
        val reg = registry(dao, resource(name = "One"), resource(name = "Two"))
        reg.install("https://one.example/m.json")
        reg.install("https://two.example/m.json")
        val before = reg.listInstalled()
        assertEquals(2, before.size)
        reg.uninstall(before[0].id)
        val after = reg.listInstalled()
        assertEquals(1, after.size)
        assertFalse(after[0].id == before[0].id)
    }

    @Test
    fun `setEnabled persists the toggle`() = runTest {
        val dao = FakeAddonDao()
        val reg = registry(dao, resource())
        reg.install("https://example.com/manifest.json")
        val id = dao.rows[0].id
        reg.setEnabled(id, false)
        assertFalse(reg.listInstalled()[0].enabled)
        reg.setEnabled(id, true)
        assertTrue(reg.listInstalled()[0].enabled)
    }

    @Test
    fun `listInstalled returns rows in install order with parsed manifests`() = runTest {
        val dao = FakeAddonDao()
        val reg = registry(dao, resource(name = "One"), resource(name = "Two"))
        reg.install("https://one.example/m.json")
        reg.install("https://two.example/m.json")
        val rows = reg.listInstalled()
        assertTrue(rows[0].addedAt <= rows[1].addedAt)
        assertEquals("One", rows[0].manifest.name)
    }

    @Test
    fun `catalogProviders appends enabled addons in install order and excludes disabled`() = runTest {
        val dao = FakeAddonDao()
        val reg = registry(dao, resource(name = "One"), resource(name = "Two"))
        reg.install("https://one.example/m.json")
        reg.install("https://two.example/m.json")
        reg.setEnabled(dao.rows[0].id, false)

        val rows = reg.listInstalled()
        val providers = catalogProvidersWithAddons(
            builtIns = listOf(FakeCatalogProvider("g", CatalogSourceKind.BUILTIN), FakeCatalogProvider("ol", CatalogSourceKind.BUILTIN)),
            curated = FakeCatalogProvider("curated", CatalogSourceKind.CURATED),
            installedAddons = rows
        )
        val sources = providers.flatMap { it.listSources() }
        assertEquals(1, sources.count { it.kind == CatalogSourceKind.ADDON })
        assertEquals(addonSource(dao.rows[1].id), sources.last { it.kind == CatalogSourceKind.ADDON }.sourceId)
        // built-ins first, curated next, addons last
        assertEquals(CatalogSourceKind.BUILTIN, sources[0].kind)
        assertEquals(CatalogSourceKind.BUILTIN, sources[1].kind)
        assertEquals(CatalogSourceKind.CURATED, sources[2].kind)
        assertEquals(CatalogSourceKind.ADDON, sources[3].kind)
    }

    @Test
    fun `catalogProviders with zero addons matches the zero-addon parity set`() {
        val providers = catalogProvidersWithAddons(
            builtIns = listOf(FakeCatalogProvider("g", CatalogSourceKind.BUILTIN), FakeCatalogProvider("ol", CatalogSourceKind.BUILTIN)),
            curated = FakeCatalogProvider("curated", CatalogSourceKind.CURATED),
            installedAddons = emptyList()
        )
        val kinds = providers.flatMap { it.listSources() }.map { it.kind }
        assertEquals(listOf(CatalogSourceKind.BUILTIN, CatalogSourceKind.BUILTIN, CatalogSourceKind.CURATED), kinds)
    }

    @Test
    fun `addon catalog provider is browse-only and exposes one addon source`() = runTest {
        val addonId = AddonId.fromUrl("https://example.com/manifest.json")
        val manifest = ManifestValidator.validate(VALID_MANIFEST_JSON.toByteArray(), "application/json")
        val provider = AddonCatalogProvider(manifest, addonId, FakeAddonHttpTransport())
        val sources = provider.listSources()
        assertEquals(1, sources.size)
        assertEquals(addonSource(addonId), sources[0].sourceId)
        assertEquals(CatalogSourceKind.ADDON, sources[0].kind)
        assertEquals("My Addon", sources[0].name)
        val page = provider.search("anything", 1)
        assertEquals(0, page.results.size)
        assertEquals(null, page.nextPage)
        assertEquals(0, page.totalCount)
        try {
            provider.getDetails("addon:$addonId:some-book")
            fail("expected NOT_FOUND")
        } catch (err: CatalogException) {
            assertEquals(CatalogErrorCode.NOT_FOUND, err.code)
        }
    }
}

File: desktop/src/test/unit/services/addons/addon-registry.test.ts
import { describe, expect, it } from 'vitest';

import {
  AddonFetchErrorCode,
  AddonRegistry,
  type AddonManifest,
} from '$lib/shared/services/addons/AddonRegistry';
import { AddonCatalogProvider } from '$lib/shared/services/addons/AddonCatalogProvider';
import { defaultCatalogProviders } from '$lib/shared/services/catalog/CompositeCatalogProvider';
import { addonIdFromUrl } from '$lib/shared/services/addons/addonId';

const VALID_MANIFEST: AddonManifest = {
  id: 'my-addon',
  name: 'My Addon',
  version: '1.0.0',
  catalogs: [{ type: 'book', id: 'main', name: 'Main catalog' }],
  resources: ['catalog'],
};

interface FakeRow {
  id: string;
  url: string;
  manifestJson: string;
  enabled: boolean;
  addedAt: number;
}

/** Mirrors the Rust/Room UPSERT semantics: reinstall preserves enabled + addedAt. */
function fakeStore() {
  const rows: FakeRow[] = [];
  return {
    rows,
    async listInstalled(): Promise<FakeRow[]> {
      return rows.map((r) => ({ ...r }));
    },
    async upsert(row: FakeRow): Promise<void> {
      const existing = rows.find((r) => r.id === row.id);
      if (existing) {
        existing.manifestJson = row.manifestJson;
      } else {
        rows.push({ ...row });
      }
    },
    async setEnabled(id: string, enabled: boolean): Promise<void> {
      const row = rows.find((r) => r.id === id);
      if (row) row.enabled = enabled;
    },
    async uninstall(id: string): Promise<void> {
      const i = rows.findIndex((r) => r.id === id);
      if (i >= 0) rows.splice(i, 1);
    },
  };
}

type TransportResult = { status: number; contentType: string | null; body: Uint8Array };

/** Transport returning a different manifest per install call, in order. */
function scriptedTransport(...manifests: unknown[]) {
  let call = 0;
  const transport = async (_url: string): Promise<TransportResult> => {
    const manifest = manifests[Math.min(call, manifests.length - 1)];
    call += 1;
    return { status: 200, contentType: 'application/json', body: jsonBytes(manifest) };
  };
  return transport;
}

function okTransport(manifest: unknown = VALID_MANIFEST, status = 200) {
  const transport = async (_url: string): Promise<TransportResult> => ({
    status,
    contentType: 'application/json',
    body: new TextEncoder().encode(JSON.stringify(manifest)),
  });
  return Object.assign(transport, { calls: 0 }) as typeof transport & { calls: number };
}

function jsonBytes(value: unknown): Uint8Array {
  return new TextEncoder().encode(JSON.stringify(value));
}

describe('AddonRegistry.install', () => {
  it('happy path stores validated manifest keyed by sha256 addonId', async () => {
    const store = fakeStore();
    const transport = okTransport();
    const registry = new AddonRegistry({ store, transport });
    const url = 'https://example.com/manifest.json';
    const manifest = await registry.install(url);
    expect(manifest.id).toBe('my-addon');
    expect(store.rows).toHaveLength(1);
    expect(store.rows[0].id).toBe(await addonIdFromUrl(url));
    expect(store.rows[0].url).toBe(url);
    expect(store.rows[0].enabled).toBe(true);
    expect(JSON.parse(store.rows[0].manifestJson).name).toBe('My Addon');
  });

  it('rejects non-HTTPS URLs before any network I/O', async () => {
    const store = fakeStore();
    const transport = okTransport();
    const registry = new AddonRegistry({ store, transport });
    await expect(registry.install('http://example.com/manifest.json')).rejects.toMatchObject({
      code: AddonFetchErrorCode.HTTPS_REQUIRED,
    });
    await expect(registry.install('ftp://example.com/manifest.json')).rejects.toMatchObject({
      code: AddonFetchErrorCode.HTTPS_REQUIRED,
    });
    expect((transport as typeof transport & { calls: number }).calls).toBe(0);
    expect(store.rows).toHaveLength(0);
  });

  it('is idempotent: reinstall updates manifest and preserves enabled', async () => {
    const store = fakeStore();
    const transport = scriptedTransport(VALID_MANIFEST, { ...VALID_MANIFEST, version: '2.0.0' });
    const registry = new AddonRegistry({ store, transport });
    const url = 'https://example.com/manifest.json';
    await registry.install(url);
    const id = store.rows[0].id;
    await registry.setEnabled(id, false);
    await registry.install(url);
    expect(store.rows).toHaveLength(1);
    expect(store.rows[0].id).toBe(id);
    expect(JSON.parse(store.rows[0].manifestJson).version).toBe('2.0.0');
    expect(store.rows[0].enabled).toBe(false);
  });

  it('rejects non-2xx fetch status as network error', async () => {
    const registry = new AddonRegistry({ store: fakeStore(), transport: okTransport(VALID_MANIFEST, 404) });
    await expect(registry.install('https://example.com/manifest.json')).rejects.toMatchObject({
      code: AddonFetchErrorCode.NETWORK,
    });
  });

  it('rejects HTML content type as bad content type', async () => {
    const transport = async (): Promise<TransportResult> => ({
      status: 200,
      contentType: 'text/html',
      body: jsonBytes('<html>not a manifest</html>'),
    });
    const registry = new AddonRegistry({ store: fakeStore(), transport });
    await expect(registry.install('https://example.com/manifest.json')).rejects.toMatchObject({
      code: AddonFetchErrorCode.BAD_CONTENT_TYPE,
    });
  });

  it('rejects oversized manifests before parse', async () => {
    const transport = async (): Promise<TransportResult> => ({
      status: 200,
      contentType: 'application/json',
      body: new Uint8Array(65 * 1024),
    });
    const registry = new AddonRegistry({ store: fakeStore(), transport });
    await expect(registry.install('https://example.com/manifest.json')).rejects.toMatchObject({
      code: AddonFetchErrorCode.TOO_LARGE,
    });
  });

  it('rejects manifests whose id collides with a built-in source name', async () => {
    const store = fakeStore();
    const registry = new AddonRegistry({ store, transport: okTransport({ ...VALID_MANIFEST, id: 'gutendex' }) });
    await expect(registry.install('https://example.com/manifest.json')).rejects.toMatchObject({
      code: AddonFetchErrorCode.INVALID_MANIFEST,
    });
    expect(store.rows).toHaveLength(0);
  });
});

describe('AddonRegistry.manage', () => {
  it('uninstall removes the row; other rows unaffected', async () => {
    const store = fakeStore();
    const registry = new AddonRegistry({ store, transport: okTransport() });
    await registry.install('https://one.example/m.json');
    await registry.install('https://two.example/m.json');
    const before = await registry.listInstalled();
    expect(before).toHaveLength(2);
    await registry.uninstall(before[0].id);
    const after = await registry.listInstalled();
    expect(after).toHaveLength(1);
    expect(after[0].id).not.toBe(before[0].id);
  });

  it('setEnabled persists the toggle', async () => {
    const store = fakeStore();
    const registry = new AddonRegistry({ store, transport: okTransport() });
    await registry.install('https://example.com/manifest.json');
    const id = store.rows[0].id;
    await registry.setEnabled(id, false);
    expect((await registry.listInstalled())[0].enabled).toBe(false);
    await registry.setEnabled(id, true);
    expect((await registry.listInstalled())[0].enabled).toBe(true);
  });

  it('listInstalled surfaces rows in install order with parsed manifests', async () => {
    const store = fakeStore();
    const registry = new AddonRegistry({ store, transport: okTransport() });
    await registry.install('https://one.example/m.json');
    await registry.install('https://two.example/m.json');
    const rows = await registry.listInstalled();
    expect(rows[0].addedAt).toBeLessThanOrEqual(rows[1].addedAt);
    expect(rows[0].manifest.name).toBe('My Addon');
  });
});

describe('AddonCatalogProvider + dynamic composite', () => {
  it('exposes one addon-kind source and is browse-only', async () => {
    const addonId = await addonIdFromUrl('https://example.com/manifest.json');
    const provider = new AddonCatalogProvider(VALID_MANIFEST, addonId);
    const sources = provider.listSources();
    expect(sources).toHaveLength(1);
    expect(sources[0].sourceId).toBe(`addon:${addonId}`);
    expect(sources[0].kind).toBe('addon');
    expect(sources[0].name).toBe('My Addon');
    await expect(provider.search('anything', 1)).resolves.toEqual({
      results: [],
      nextPage: null,
      totalCount: 0,
    });
    await expect(provider.getDetails(`addon:${addonId}:some-book`)).rejects.toMatchObject({
      code: 'NOT_FOUND',
    });
  });

  it('defaultCatalogProviders appends enabled addons in install order and excludes disabled', async () => {
    const store = fakeStore();
    const registry = new AddonRegistry({ store, transport: okTransport() });
    await registry.install('https://one.example/m.json');
    await registry.install('https://two.example/m.json');
    await registry.setEnabled(store.rows[0].id, false);

    const rows = await registry.listInstalled();
    const providers = defaultCatalogProviders(rows);
    const sources = providers.flatMap((p) => p.listSources());
    expect(sources.filter((s) => s.kind === 'addon')).toHaveLength(1);
    expect(sources.filter((s) => s.kind === 'addon')[0].sourceId).toBe(`addon:${store.rows[1].id}`);
    // built-ins first, curated next (6 sources), addons last
    expect(sources[0].kind).toBe('builtin');
    expect(sources[1].kind).toBe('builtin');
    expect(sources[2].kind).toBe('curated');
    expect(sources[7].kind).toBe('curated');
    expect(sources[8].kind).toBe('addon');
  });

  it('defaultCatalogProviders with zero addons matches the zero-addon parity set', () => {
    const providers = defaultCatalogProviders([]);
    const kinds = providers.flatMap((p) => p.listSources()).map((s) => s.kind);
    expect(kinds).toEqual(['builtin', 'builtin', 'curated', 'curated', 'curated', 'curated', 'curated', 'curated']);
  });
});
